kmip/tests/unit/services/server/crypto/test_engine.py

```diff
@@ -1331,6 +1331,34 @@ def test_verify_signature_invalid_signing_algorithm(self):
              b'\x6a\xcc\x04\x14\x2e\x10\x0a\x65'
              b'\xf5\x1b\x97\xad\xf5\x17\x2c\x41'
          )},
+        {'algorithm': enums.CryptographicAlgorithm.AES,
+         'cipher_mode': enums.BlockCipherMode.GCM,
+         'key': (
+             b'\x6e\xd7\x6d\x2d\x97\xc6\x9f\xd1'
+             b'\x33\x95\x89\x52\x39\x31\xf2\xa6'
+             b'\xcf\xf5\x54\xb1\x5f\x73\x8f\x21'
+             b'\xec\x72\xdd\x97\xa7\x33\x09\x07'
+         ),
+         'iv_nonce': (
+             b'\x85\x1e\x87\x64\x77\x6e\x67\x96'
+             b'\xaa\xb7\x22\xdb\xb6\x44\xac\xe8'
+         ),
+         'plain_text': (
+             b'\x62\x82\xb8\xc0\x5c\x5c\x15\x30'
+             b'\xb9\x7d\x48\x16\xca\x43\x47\x62'
+         ),
+         'auth_additional_data': (
+             b'\x61\x75\x74\x68\x2d\x74\x65\x73\x74'
+         ),
+         'auth_tag_length': 16,
+         'cipher_text': (
+             b'\x09\x45\x1b\x14\xf9\x25\x68\x24'
+             b'\x55\x50\x13\xb5\x09\xe9\xd6\x63'
+         ),
+         'auth_tag': (
+             b'\xbe\x9f\xef\x78\x5f\x9e\xe6\x16'
+             b'\x90\xe9\x44\x59\xc1\x84\x44\x7f'
+         )},
```

Comment:
Where does this test vector come from?

Reply:
I just ran PyKMIP encode in GCM mode with pre-defined nonce, got this value as a result and recorded it.

Reply:
That's what I thought. For this we need to test with known valid GCM test vectors, to prove that encrypt and decrypt work correctly with GCM. Many of the other test vectors in `test_engine.py` are pulled from existing NIST or IETF sources (see [here](https://github.com/OpenKMIP/PyKMIP/blob/b7a4e9be120185af1e3c819797416cd5c8b082c9/kmip/tests/unit/services/server/crypto/test_engine.py#L1786)). GCM test vectors should be available from NIST as well (see [here](https://csrc.nist.gov/Projects/cryptographic-algorithm-validation-program/CAVP-TESTING-BLOCK-CIPHER-MODES#GCMVS)), specifically the [Validation System document](https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Algorithm-Validation-Program/documents/mac/gcmvs.pdf) and the [test vector archive](http://csrc.nist.gov/groups/STM/cavp/documents/mac/gcmtestvectors.zip). It looks like [pyca/cryptography](https://github.com/pyca/cryptography) also has a [copy](https://github.com/pyca/cryptography/tree/master/vectors/cryptography_vectors/ciphers/AES/GCM) of the test vector archive, but I'd pull directly from the original source.

Take a look at how other tests handle the test vectors and pull in one or more test vectors from the NIST archive, replacing the test vector you have above. Be sure to add a comment pointing specifically to where you got the test vector. Once that's in place and the tests all pass, this should be good to merge.

Reply:
Thanks, this test case looks good to me.